php/class-wp-app-manifest.php

```diff
@@ -0,0 +1,209 @@
+	 *
+	 * Copied from Jetpack_PWA_Helpers::get_default_manifest_icon_sizes().
+	 * Based on a conversation in https://github.com/GoogleChrome/lighthouse/issues/291
+	 *
+	 * @var int[]
+	 */
+	public $default_manifest_icon_sizes = array( 192, 512 );
+
+	/**
+	 * Inits the class.
+	 *
+	 * Mainly copied from Jetpack_PWA_Manifest::__construct().
+	 */
+	public function init() {
+		add_action( 'wp_head', array( $this, 'manifest_link_and_meta' ) );
+		add_action( 'rest_api_init', array( $this, 'register_manifest_rest_route' ) );
+	}
+
+	/**
+	 * Outputs the <link> and <meta> tags for the app manifest.
+	 *
+	 * Mainly copied from Jetpack_PWA_Manifest::render_manifest_link().
+	 */
+	public function manifest_link_and_meta() {
+		?>
+		<link rel="manifest" href="<?php echo esc_url( rest_url( self::REST_NAMESPACE . self::REST_ROUTE ) ); ?>">
+		<meta name="theme-color" content="<?php echo esc_attr( $this->get_theme_color() ); ?>">
+		<?php
+	}
+
+	/**
+	 * Gets the theme color for the manifest.
+	 *
+	 * Mainly copied from Jetpack_PWA_Helpers::get_theme_color().
+	 * First looks for the header background color in the AMP for WordPress plugin, if it's active.
+	 * This color displays on loading the app.
+	 *
+	 * @return string $theme_color The theme color for the manifest.json file, as a hex value.
+	 */
+	public function get_theme_color() {
+		if ( current_theme_supports( 'custom-background' ) ) {
+			$background_color = get_background_color(); // This returns a hex value without the leading #, or an empty string.
+			if ( $background_color ) {
+				$theme_color = "#$background_color";
+			}
+		}
+
+		if ( ! isset( $theme_color ) ) {
+			$theme_color = self::FALLBACK_THEME_COLOR;
+		}
+
+		/**
+		 * Enables overriding the PWA theme color.
+		 *
+		 * @param string $theme_color Hex color value.
+		 */
+		return apply_filters( 'pwa_background_color', $theme_color );
+	}
+
+	/**
+	 * Gets the manifest data for the REST API response.
+	 *
+	 * Mainly copied from Jetpack_PWA_Helpers::render_manifest_json().
+	 */
+	public function get_manifest() {
+		$manifest = array(
+			'name'      => get_bloginfo( 'name' ),
+			'start_url' => get_home_url(),
+			'display'   => 'minimal-ui',
+			'lang'      => get_locale(),
+			'dir'       => is_rtl() ? 'rtl' : 'ltr',
+		);
+
+		/**
+		 * Gets the 'short_name' by limiting the blog name to 12 characters.
+		 * And if this cuts off a word, it omits the word entirely by using a positive look-ahead in the regex.
+		 * For example, the first 12 characters of 'My PWA WordPress Site' are 'My PWA WordP'.
+		 * Because this cuts off the last word, this removes 'WordP' entirely: 'My PWA'.
+		 *
+		 * @link https://stackoverflow.com/questions/12646197/cut-the-string-to-be-80-characters-and-must-keep-the-words-without-cutting-th#answer-12646400
+		 */
+		preg_match( '/^.{0,12}(?= |$)/', get_bloginfo( 'name' ), $short_name_matches );
+		if ( $short_name_matches ) {
+			$manifest['short_name'] = $short_name_matches[0];
```

Comment:
The `short_name` might need more thought. This limits it to the maximum of 12 characters that [Chrome recommends](https://developer.chrome.com/apps/manifest/name#short_name). But this could still be awkward.